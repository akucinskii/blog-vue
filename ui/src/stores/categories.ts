import { defineStore } from "pinia";
import { CategoryInterface } from "../types/types";
import apiRequest, { HTTP_METHODS } from "../utils/apiRequest";

export interface RootState {
  categories: CategoryInterface[];
  //selectedCategory: CategoryInterface | null;
}

export const useCategoriesStore = defineStore("categories", {
  state: () =>
    ({
      categories: [],
    } as RootState),
  getters: {
    getCategoryByUUID: (state) => {
      return (categoryId: string) => state.categories.find((category) => category.id === categoryId);
    },
  },
  actions: {
    async fetchCategories() {
      try {
        this.categories = await apiRequest(HTTP_METHODS.GET, "/categories?page=1&size=50").then(
          (response) => response.data.items
        );
      } catch (e) {
        this.categories = [];
      }
    },
  },
});
