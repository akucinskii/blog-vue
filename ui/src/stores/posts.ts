import { defineStore } from "pinia";
import { ArticleInterface } from "../types/types";
import apiRequest from "../utils/apiRequest";
import { HTTP_METHODS } from "../utils/apiRequest";
export interface RootState {
  posts: ArticleInterface[];
  loading: boolean;
  error: null | unknown;
}

export const usePostsStore = defineStore("posts", {
  state: () =>
    ({
      posts: [],
      loading: false,
      error: null,
    } as RootState),
  getters: {
    getPostByUUID: (state) => {
      return (postId: string) => state.posts.find((post) => post.id === postId);
    },
    getPostsByCategory: (state) => {
      return (category: string) => state.posts.filter((post) => post.category.id === category);
    },
  },
  actions: {
    async fetchPosts() {
      this.loading = true;
      try {
        this.posts = await apiRequest(HTTP_METHODS.GET, "/posts?page=1&size=50").then(
          (response) => response.data.items
        );
      } catch (e) {
        this.error = e;
        this.posts = [];
      }

      this.loading = false;
    },
    async removePost(postId: string) {
      await apiRequest(HTTP_METHODS.DELETE, `/posts?post_id=${postId}`);
    },
  },
});
