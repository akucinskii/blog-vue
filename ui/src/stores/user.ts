import { defineStore } from "pinia";
import apiRequest, { HTTP_METHODS } from "../utils/apiRequest";

export type RootState = {
  user: {
    username: string;
    avatar_url: string;
    id: string;
    description: string;
    name: string;
    surname: string;
    is_admin: boolean;
  } | null;
};

export const useUserStore = defineStore("user", {
  state: () =>
    ({
      user: null,
    } as RootState),
  getters: {},
  actions: {
    async fetchUser() {
      try {
        const result = await apiRequest(HTTP_METHODS.GET, "/me", {});
        this.user = result.data;
      } catch (e) {
        this.user = null;
      }
    },

    logOut() {
      localStorage.removeItem("token");
      this.user = null;
    },
  },
});
