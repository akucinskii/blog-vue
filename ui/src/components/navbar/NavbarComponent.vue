<script setup lang="ts">
import NavbarLogo from "./NavbarLogo.vue";
import NavbarSocialMedia from "./NavbarSocialMedia.vue";
import NavbarSearch from "./NavbarSearch.vue";
import { onMounted, onUnmounted, ref, watch } from "vue";
import checkIsMobile from "../../utils/checkIsMobile";
import { useCategoriesStore } from "../../stores/categories";
import { storeToRefs } from "pinia";
import { useUserStore } from "../../stores/user";
import { useRouter } from "vue-router";

const store = useCategoriesStore();
const userStore = useUserStore();
const router = useRouter();

const { categories } = storeToRefs(store);
const { user } = storeToRefs(userStore);
const drawer = ref(false);
const username = ref("");
const isMobile = ref(true);

const resizeHandler = () => {
  isMobile.value = checkIsMobile();
};

onMounted(() => {
  window.addEventListener("resize", resizeHandler);
  isMobile.value = checkIsMobile();
  if (categories.value.length === 0) {
    store.fetchCategories();
  }
});

watch(user, () => {
  username.value = user.value ? user.value.username : "";
});

onUnmounted(() => {
  window.removeEventListener("resize", resizeHandler);
});

defineExpose({
  username,
});
</script>

<template>
  <v-app-bar v-if="!isMobile" id="desktop">
    <v-spacer />
    <v-row no-gutters>
      <v-col class="navbar__container--small">
        <NavbarLogo />
      </v-col>
      <v-col class="navbar__container--large">
        <router-link v-for="category in categories" :key="category.id" :to="`/category/${category.id}`">
          <v-btn variant="text">
            {{ category.name }}
          </v-btn>
        </router-link>
      </v-col>
      <v-col class="navbar__container--small">
        <NavbarSearch />
        <NavbarSocialMedia />
        <div v-if="!username" class="d-flex">
          <router-link to="/login">
            <v-btn variant="tonal" class="mx-2" color="red"> Login </v-btn>
          </router-link>
          <router-link to="/register">
            <v-btn variant="plain"> Register </v-btn>
          </router-link>
        </div>
        <div v-else>
          <v-btn @click="router.push('/profile')">{{ username }}</v-btn>
        </div>
      </v-col>
    </v-row>

    <v-spacer />
  </v-app-bar>

  <v-app-bar v-else id="mobile">
    <v-app-bar-nav-icon id="hamburgerMenu" variant="text" @click="drawer = !drawer"></v-app-bar-nav-icon>
    <v-spacer />

    <NavbarLogo />

    <v-spacer />

    <NavbarSocialMedia />
  </v-app-bar>

  <v-navigation-drawer v-model="drawer" class="py-4 m-0 text-center" bottom temporary>
    <v-list class="d-flex align-center flex-column"
      ><v-list-item v-for="category in categories" :key="category.id">
        <router-link :to="`/category/${category.id}`">
          <v-btn variant="text">
            {{ category.name }}
          </v-btn>
        </router-link>
      </v-list-item>
      <v-list-item v-if="!username">
        <router-link to="/login">
          <v-btn variant="tonal" class="mx-2" color="red"> Login </v-btn>
        </router-link>
      </v-list-item>
      <v-list-item v-if="!username">
        <router-link to="/register">
          <v-btn variant="plain"> Register </v-btn>
        </router-link>
      </v-list-item>

      <v-list-item v-else>
        <v-btn class="text--white" color="red" @click="router.push('/profile')">{{ username }}</v-btn>
      </v-list-item>
    </v-list>
  </v-navigation-drawer>
</template>

<style lang="scss" scoped>
.navbar__container--large {
  display: flex;
  align-items: center;
  gap: 1rem;
}
.navbar__container--small {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 20%;
}
.text--white {
  color: white !important;
}
</style>
