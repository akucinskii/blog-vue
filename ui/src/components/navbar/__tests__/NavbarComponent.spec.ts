import { createTestingPinia } from "@pinia/testing";
import { shallowMount, mount } from "@vue/test-utils";
import { setActivePinia } from "pinia";
import { nextTick } from "vue";
import { useUserStore } from "../../../stores/user";
import NavbarComponent from "../NavbarComponent.vue";

describe("NavbarComponent", () => {
  beforeEach(() => {
    window.innerWidth = 20;
    setActivePinia(
      createTestingPinia({
        initialState: {
          categories: {
            categories: [],
            user: {},
          },
        },
      })
    );

    jest.restoreAllMocks();
  });

  test("creates navbar", () => {
    const wrapper = shallowMount(NavbarComponent);

    expect(wrapper.find("v-app-bar").exists()).toBe(true);
  });

  test("Creates a clickable button on mobile", async () => {
    const wrapper = shallowMount(NavbarComponent);

    expect(wrapper.find("#hamburgerMenu").exists()).toBe(true);
  });

  test("There is no drawer on desktop", () => {
    window.innerWidth = 1265;
    const wrapper = shallowMount(NavbarComponent);

    expect(wrapper.find("#v-navigation-drawer").exists()).toBe(false);
  });

  test("onClick shows drawer on mobile", async () => {
    const wrapper = shallowMount(NavbarComponent);

    await wrapper.find("#hamburgerMenu").trigger("click");

    expect(wrapper.find("v-navigation-drawer").exists()).toBe(true);
  });

  test("It should match Snapshot", () => {
    const wrapper = mount(NavbarComponent);

    expect(wrapper.element).toMatchSnapshot();
  });

  test("Should create/remove event listeners", () => {
    const addEventListenerSpy = jest.spyOn(window, "addEventListener");
    const removeEventListenerSpy = jest.spyOn(window, "removeEventListener");
    const wrapper = shallowMount(NavbarComponent);

    expect(addEventListenerSpy).toHaveBeenCalledWith("resize", expect.any(Function));

    wrapper.unmount();

    expect(removeEventListenerSpy).toHaveBeenCalledWith("resize", expect.any(Function));
  });

  test("Should show correct object based on device width", async () => {
    window.innerWidth = 1;
    const wrapper = mount(NavbarComponent);

    expect(wrapper.find("#mobile").exists()).toBe(true);
    expect(wrapper.find("#desktop").exists()).toBe(false);

    window.innerWidth = 1500;
    window.dispatchEvent(new Event("resize"));
    await nextTick();

    expect(wrapper.find("#mobile").exists()).toBe(false);
    expect(wrapper.find("#desktop").exists()).toBe(true);
  });

  test("username changes correctly", async () => {
    const wrapper = mount(NavbarComponent);

    const userStore = useUserStore();

    expect(wrapper.vm.username).toEqual("");

    userStore.user = {
      username: "John",
      avatar_url: "test",
      id: "1",
      description: "test",
      surname: "test",
      is_admin: false,
      name: "John",
    };
    await nextTick();

    expect(wrapper.vm.username).toEqual("John");
    expect(wrapper.vm.username).toEqual(userStore.user.username);

    userStore.user = null;

    await nextTick();

    expect(wrapper.vm.username).toEqual("");
  });
});
