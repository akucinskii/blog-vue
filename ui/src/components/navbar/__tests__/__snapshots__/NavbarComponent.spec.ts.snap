// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavbarComponent It should match Snapshot 1`] = `
<div
  data-v-app=""
>
  
  <v-app-bar
    id="mobile"
  >
    <v-app-bar-nav-icon
      id="hamburgerMenu"
      variant="text"
    />
    <v-spacer />
    <router-link
      to="/"
    >
      Dieta by Robercik
    </router-link>
    <v-spacer />
    <div
      class="social-media__container"
    >
      <a
        href="https://facebook.com/rl9official"
        target="_blank"
      >
        <button>
          <v-icon>
            mdi-facebook
          </v-icon>
        </button>
      </a>
      <a
        href="https://www.youtube.com/channel/UCZXD4PIYIwC48tkTsdGAKYw"
        target="_blank"
      >
        <button>
          <v-icon>
            mdi-youtube
          </v-icon>
        </button>
      </a>
      <a
        href="https://www.instagram.com/_rl9"
        target="_blank"
      >
        <button>
          <v-icon>
            mdi-instagram
          </v-icon>
        </button>
      </a>
    </div>
  </v-app-bar>
  <v-navigation-drawer
    bottom=""
    class="py-4 m-0 text-center"
    modelvalue="false"
    temporary=""
  >
    <v-list
      class="d-flex align-center flex-column"
    >
      
      
      <v-list-item>
        <router-link
          to="/login"
        >
          <v-btn
            class="mx-2"
            color="red"
            variant="tonal"
          >
             Login 
          </v-btn>
        </router-link>
      </v-list-item>
      <v-list-item>
        <router-link
          to="/register"
        >
          <v-btn
            variant="plain"
          >
             Register 
          </v-btn>
        </router-link>
      </v-list-item>
    </v-list>
  </v-navigation-drawer>
  
</div>
`;
